### PR TITLE
Enable effects v2 in protocol config

### PR DESCRIPTION
**Pull request by lxfind**
## Description 

This PR enables transaction effects v2 in any non-mainnet network.
TODO: Decide the right protocol version to enable it before merging.

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-types/src/effects/mod.rs [hunk 1]
```diff
@@ -388,6 +329,7 @@ pub trait TransactionEffectsAPI {
     fn into_status(self) -> ExecutionStatus;
     fn executed_epoch(&self) -> EpochId;
     fn modified_at_versions(&self) -> Vec<(ObjectID, SequenceNumber)>;
+    fn modified_at_v2(&self) -> Vec<(ObjectRef, Owner)>;
```
Comment (longbowlu) [inline crates/sui-types/src/effects/mod.rs:332]: is this "after modification" or "before"? can we name it more clearly

File: crates/sui-core/src/authority.rs [hunk 2]
```diff
@@ -1403,7 +1403,7 @@ impl AuthorityState {
         let gas_status =
             SuiGasStatus::new(max_tx_gas, gas_price, reference_gas_price, protocol_config)?;
 
-        let gas_object_id = ObjectID::random();
+        let gas_object_id = ObjectID::MAX;
```
Comment (mystenmark) [inline crates/sui-core/src/authority.rs:1406]: why was this necessary?

File: crates/sui-types/src/effects/effects_v2.rs [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+    status: ExecutionStatus,
+    /// The epoch when this transaction was executed.
+    executed_epoch: EpochId,
+    gas_used: GasCostSummary,
+    /// The transaction digest
+    transaction_digest: TransactionDigest,
+    /// The updated gas object reference, as an index into the `changed_objects` vector.
+    /// Having a dedicated field for convenient access.
+    /// System transaction that don't require gas will leave this as None.
+    gas_object_index: Option<u16>,
```
Comment (mystenmark) [inline crates/sui-types/src/effects/effects_v2.rs:33]: I don't think its completely inconceivable we would want to exceed 2^16 outputs at some point (due to child objects), and given that this is a huge struct already (probably 200-300 bytes from eyeballing it), why not just make this a `u32` or even `usize`?

File: crates/sui-types/src/effects/effects_v2.rs [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+
+    fn into_status(self) -> ExecutionStatus {
+        self.status
+    }
+
+    fn executed_epoch(&self) -> EpochId {
+        self.executed_epoch
+    }
+
+    // TODO: Add a new API to return modified object refs.
```
Comment (mystenmark) [inline crates/sui-types/src/effects/effects_v2.rs:68]: this TODO should probably be in the trait def, not the impl

File: crates/sui-types/src/effects/effects_v2.rs [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+        if let Some(gas_object_index) = self.gas_object_index {
+            let entry = &self.changed_objects[gas_object_index as usize];
+            match entry.1.output_state {
+                ObjectOut::ObjectWrite((digest, owner)) => {
+                    ((entry.0, self.lamport_version, digest), owner)
+                }
+                _ => panic!("Gas object must be an ObjectWrite in changed_objects"),
+            }
+        } else {
+            // TODO: We should consider having this function to return Option.
```
Comment (mystenmark) [inline crates/sui-types/src/effects/effects_v2.rs:246]: ditto

File: crates/sui-types/src/effects/effects_v2.rs [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+        Self {
+            transaction_digest: tx.digest(),
+            status,
+            ..Default::default()
+        }
+    }
+
+    /// This function demonstrates what's the invariant of the effects.
+    /// It also documents the semantics of different combinations in object changes.
+    /// TODO: Make this debug assertion only.
```
Comment (mystenmark) [inline crates/sui-types/src/effects/effects_v2.rs:413]: looks like this TODO is done?

File: crates/sui-types/src/execution.rs [hunk 7]
```diff
@@ -101,6 +99,72 @@ pub struct ExecutionResultsV2 {
...
+        self.user_events.extend(new_results.user_events);
+    }
+
+    pub fn update_version_and_previous_tx(
+        &mut self,
+        lamport_version: SequenceNumber,
+        prev_tx: TransactionDigest,
+    ) {
+        for (id, mut obj) in self.written_objects.iter_mut() {
+            // TODO: All of the following is no longer necessary and can be simplified.
```
Comment (mystenmark) [inline crates/sui-types/src/execution.rs:127]: if this todo is still applicable can you make it more clear what the simplification would be?

**Review by mystenmark — COMMENTED** [review]
How are all the temporary store changes tested / gated?

File: crates/sui-json-rpc-types/src/sui_transaction.rs [hunk 8]
```diff
@@ -638,56 +638,43 @@ impl TryFrom<TransactionEffects> for SuiTransactionBlockEffects {
...
-            .expect("TransactionEffects defines message_version()");
-
-        match message_version {
-            1 => Ok(SuiTransactionBlockEffects::V1(
-                SuiTransactionBlockEffectsV1 {
-                    status: effect.status().clone().into(),
-                    executed_epoch: effect.executed_epoch(),
-                    modified_at_versions: effect
-                        .modified_at_versions()
+        Ok(SuiTransactionBlockEffects::V1(
```
[the remark this answers is not in the document]
Comment (lxfind) [reply crates/sui-json-rpc-types/src/sui_transaction.rs:641]: Because we don't have SuiTransactionBlockEffectsV2 yet. It's not clear to me what are the steps involved to add SuiTransactionBlockEffectsV2 to the RPC layer. But I think we could ship this to core first and then look at RPC layer. I am also curious how this overlaps with the graphql timeline.

File: crates/sui-types/src/effects/mod.rs [hunk 9]
```diff
@@ -53,20 +50,24 @@ pub const APPROX_SIZE_OF_OWNER: usize = 48;
...
         Some(match self {
             Self::V1(_) => 1,
+            Self::V2(_) => 2,
         })
     }
 
     fn check_version_supported(&self, protocol_config: &ProtocolConfig) -> SuiResult {
         let (message_version, supported) = match self {
             Self::V1(_) => (1, SupportedProtocolVersions::new_for_message(1, u64::MAX)),
+            Self::V2(_) => (2, SupportedProtocolVersions::new_for_message(23, u64::MAX)),
```
Comment (longbowlu) [inline crates/sui-types/src/effects/mod.rs:70]: just to make sure version 23 is what you are targeting (and not a placeholder)

File: crates/sui-json-rpc-types/src/sui_transaction.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (longbowlu) [reply crates/sui-json-rpc-types/src/sui_transaction.rs:641]: I'd hope SDK would hide these stuff
